modules/frontend/frontend.go

```diff
@@ -0,0 +1,227 @@
+}
+
+type Middleware interface {
+	Wrap(Handler) Handler
+}
+
+// MiddlewareFunc is like http.HandlerFunc, but for Middleware.
+type MiddlewareFunc func(Handler) Handler
+
+// Wrap implements Middleware.
+func (q MiddlewareFunc) Wrap(h Handler) Handler {
+	return q(h)
+}
+
+func MergeMiddlewares(middleware ...Middleware) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		for i := len(middleware) - 1; i >= 0; i-- {
+			next = middleware[i].Wrap(next)
+		}
+		return next
+	})
+}
+
+type roundTripper struct {
+	next    http.RoundTripper
+	handler Handler
+}
+
+// NewRoundTripper merges a set of middlewares into an handler, then inject it into the `next` roundtripper
+func NewRoundTripper(next http.RoundTripper, middlewares ...Middleware) http.RoundTripper {
+	transport := roundTripper{
+		next:  next,
+	}
+	transport.handler = MergeMiddlewares(middlewares...).Wrap(&transport)
+	return transport
+}
+
+func (q roundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	level.Info(util.Logger).Log("blockStart", r.URL.Query().Get("blockStart"), "blockEnd", r.URL.Query().Get("blockEnd"))
+	return q.handler.Do(r)
+}
+
+// Do implements Handler.
+func (q roundTripper) Do(r *http.Request) (*http.Response, error) {
+	level.Info(util.Logger).Log("msg", "roundTripper.Do called")
+	return q.next.RoundTrip(r)
+}
+
+func ShardingWare(shardNum int, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardQuery{
+			next:     next,
+			shardNum: shardNum,
+			logger:   logger,
+			splitByCounter: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+				Namespace: "tempo",
+				Name:      "query_frontend_split_queries_total",
+				Help:      "Total number of underlying query requests after sharding",
+			}, []string{"user"}),
+		}
+	})
+}
+
+type shardQuery struct {
+	next     Handler
+	shardNum int
+	logger   log.Logger
+	// Metrics.
+	splitByCounter *prometheus.CounterVec
+}
+
+// Do implements Handler
+func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
+	level.Info(s.logger).Log("msg", "shardQuery called")
+
+	userID, err := user.ExtractOrgID(r.Context())
+	if err != nil {
+		return nil, err
+	}
+
+	// create sharded queries
+	boundaryBytes := make([][]byte, s.shardNum+1)
+	for i := 0; i < s.shardNum+1; i ++ {
+		boundaryBytes[i] = make([]byte, 0)
+	}
+	const MaxUint = ^uint64(0)
+	const MaxInt = int64(MaxUint >> 1)
+	for i := 0; i < s.shardNum; i++ {
+		binary.PutVarint(boundaryBytes[i], MaxInt*(int64(i))/int64(s.shardNum))
+		binary.PutVarint(boundaryBytes[i], 0)
+	}
+	binary.PutVarint(boundaryBytes[s.shardNum], MaxInt)
+	binary.PutVarint(boundaryBytes[s.shardNum], MaxInt)
+
+	reqs := make([]*http.Request, s.shardNum)
+	for i := 0; i < s.shardNum; i++ {
+		reqs[i] = r
+		reqs[i].URL.Query().Add("blockStart", string(boundaryBytes[i]))
+		reqs[i].URL.Query().Add("blockEnd", string(boundaryBytes[i+1]))
```

Comment:
you need to add `queryIngesters=false` to all of these queries and then create on additional where `queryIngesters=true`.